keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -132,48 +133,97 @@ Other identified non-goals are:
+  (e.g.  Java process needs to change its Xmx flag) By using ResizePolicy, user
+  can mark Containers as safe (or unsafe) for in-place resource update. Kubelet
+  uses it to determine the required action.
+
+Note: `RestartNotRequired` does not guarantee that container won't be
+restarted. The runtime may choose to stop the container if it is unable to
+apply the new resources without doing so.
 
 Setting the flag to separately control CPU & memory is due to an observation
 that usually CPU can be added/removed without much problem whereas changes to
 available memory are more probable to require restarts.
 
-If more than one resource type with different policies are updated, then
-RestartContainer policy takes precedence over NoRestart policy.
+If more than one resource type with different policies are updated at the same
+time, then any `Restart` policy takes precedence over `RestartNotRequired` policies.
+
+If a container's RestartPolicy is `Never`, the ResizePolicy must be set to
+`RestartNotRequired` to pass validation.  That said, any in-place resize may
+result in the container being stopped *and not restarted*, if the system can
+not perform the resize in place.
+
+#### Resize Status
+
+In addition to the above, a new field `Pod.Status.ContainerStatuses[i].Resize`
+will be added.  This field indicates whether kubelet has accepted or rejected a
+proposed resize operation.  Any time the
+`Pod.Spec.Containers[i].Resources.Requests` field differs from the
+`Pod.Status.ContainerStatuses[i].Resources` field, this new field explains why.
+
+This field can be set to one of the following values:
+* Unknown - the proposed resize (in Spec...Resources) has not been accepted or
+  rejected yet.
+* InProgress - the proposed resize has been accepted and is being actuated.
+* Infeasible - the proposed resize is not feasible and is rejected.
+* (no value) - there is no proposed resize
 
-Additionally, if RestartPolicy is 'Never', ResizePolicy should be set to
-NoRestart in order to pass validation.
+Any time the apiserver observes a proposed resize (a modification of a
+`Spec...Resources` field), it will automatically set this field to "Unknown".
+
+In discussion of this KEP, the idea was proposed to distinguish "not feasible
+right now" from "not feasible ever" (i.e. "the requested resources exist on this
+node, but are busy" vs "the requested resources do not exist").  For alpha, it
+was agreed to defer this distinction.  To make this field future-safe,
+consumers should assum that any unknown value means the same as "Infeasible".
+
+If the demand arises, during alpha or after, for finer-grained information,
+this field could either be changed to or augmented by a map of resource-name to
+status.
+
+<<[UNRESOLVED]>>
+If we have this field, do we really need Status...ResourcesAllocated?  It's
+always going to be the same as either Spec...Resources or Status...Resources
+and the Resize field indicates which.  ResourcesAllocated seems redundant.
```

Comment:
We need this so scheduler can use max(Spec.C[i].Resources, Status.CS[i].ResourcesAllocated) for ranking the node. (I'm still wondering if we can just use ResourcesAllocated instead of max, despite a small possibility of race)

Updating Status.Resources.Requests & Limits could take some time as it has to go through **UpdateContainerResources** CRI API (except for memory.requests, which can be updated as soon as pod resize passes the node's canAdmit(resized)Pod test). The runtime could also fail the CRI API, in which case we would keep retrying.

So even if we accept a resize based on request fits node predicate, and signal back acceptance quickly via Pod.Status.Resize=InProgress (assuming I didn't miss something above and we agree to move Resize field to Pod level), we cannot update Status.CS[i].Resources until it successfully goes through CRI for each container in the pod. So we cannot guarantee this will converge quickly or converge at all. Scheduler won't be able to use either Spec.Resources (user could have updated it) or Status.Resources (not converged yet) during this time.

Reply:
>  I right that the ResourcesAllocated field will always be the same value as either spec or status, and never in-between? If so, I assume it would ALWAYS be something like:

ResourcesAllocated can bear an intermediate value as the user can change spec via /resize at any time as [detailed above at T=3](https://github.com/kubernetes/enhancements/pull/1883#discussion_r554660488). If we don't expose this in API, kubelet can set 'InProgress' patch qualifying it with the spec it evaluated. Perhaps it might work as long as we are ok with max(spec, status) potentially not converging for long periods. We have to be sure we can resolve competing state between Proposed from user-action & InProgress/Deferred/Infeasible. Let me think through this..

Reply:
> Am I right that the ResourcesAllocated field will always be the same value as either spec or status, and never in-between? If so, I assume it would ALWAYS be something like:

I believe you’re right. We may be able to do without exposing resourcesAllocated in the API despite the possibility that we can have spec.requests != resourcesAllocated != status.requests. 

What we lose is the ability to know what the node has actually reserved for the pod when Resize == 'Deferred' or 'Infeasible', and the impact of un-converged max(spec.resources, status.resources), particularly on scheduling. **This worries me somewhat**.. but the good thing is if we run into real-world problems, we can always just add it later and change scheduler / RQ to always use resourcesAllocated instead of max as we do in the current implementation.

Please take a look at the following. Without exposing resourcesAllocated, the [above time-sequence](https://github.com/kubernetes/enhancements/pull/1883/#discussion_r554660488) changes to:
```
T=0: New pod enters running state, kubelet checkpoints requests[cpu] = 1, starts containers, updates status.
    spec.containers[0].resources.requests[cpu] = 1
    status.Resize[cpu] = ''  (no value)
    status.containerStatuses[0].resources.requests[cpu] = 1
T=1: User update #1 requesting cpu = 1.5 via /resize subresource. apiserver patches pod object, sets Resize to ‘Proposed’.
    spec.containers[0].resources.requests[cpu] = 1.5
    status.Resize[cpu] = 'Proposed'.   <-- apiserver sets this because spec.resources != status.resources   
    status.containerStatuses[0].resources.requests[cpu] = 1
T=2: Kubelet sees update #1, accepts it, updates internal checkpoint cpu = 1.5, sends patch {spec.containers[0].resources.requests[cpu] = 1.5, status.Resize = InProgress}, and then invokes CRI.
    spec.containers[0].resources.requests[cpu] = 1.5
    status.Resize[cpu] = 'InProgress'   <-- apiserver updates it to ‘InProgress’ as {for-atleast-one-ctr spec.resources[cpu] != status.resources[cpu]} and spec in patch matches pod spec.
    status.containerStatuses[0].resources.requests[cpu] = 1
T=3: User update #2 requesting cpu = 2 via /resize, apiserver patches pod object, sets Resize to ‘Proposed’.
    spec.containers[0].resources.requests[cpu] = 2
    status.Resize[cpu] = 'Proposed'.   <-- apiserver sets this because spec.resources was updated, and spec.resources != status.resources
    status.containerStatuses[0].resources.requests[cpu] = 1
T=4: Container runtime effected cpu=1.5, CRI returns success, kubelet patches status.resources[cpu] = 1.5
    spec.containers[0].resources.requests[cpu] = 2
    status.Resize[cpu] = 'Proposed’.  <-- Stays at current value (‘Proposed’) as {for atleast one container spec.resources[cpu] != status.resources[cpu]}
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=5: No-op
T=6: Kubelet sees update #2, possible but not right now, sends patch {spec.containers[0].resources.requests[cpu] = 2, status.Resize = Deferred}
    spec.containers[0].resources.requests[cpu] = 2
    status.Resize[cpu] = 'Deferred' <-- apiserver updates it to ‘Deferred’ as {for-atleast-one-ctr spec.resources[cpu] != status.resources[cpu]} and spec in patch matches pod spec.
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=7: User sees 'Deferred' , and user policy allows user to wait for pod resize at a later time, so user waits.
T=8: Another pod leaves, kubelet can now take update #2, updates internal checkpoint cpu = 2, sends patch {spec.containers[0].resources.requests[cpu] = 2, status.Resize = InProgress}, and invokes CRI
    spec.containers[0].resources.requests[cpu] = 2
    status.Resize[cpu] = 'InProgress'   <-- apiserver updates it to ‘InProgress’ as {for-atleast-one-ctr spec.resources[cpu] != status.resources[cpu]} and spec in patch matches pod spec.
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=9: Container runtime applied cpu=2, CRI returns success, kubelet patches status.resources[cpu]    
    spec.containers[0].resources.requests[cpu] = 2
    status.Resize[cpu] = ''   <-- apiserver clears this because foreach c spec.resources[cpu] == status.resources[cpu]
    status.containerStatuses[0].resources.requests[cpu] = 2
```

Again, if at T=5 a new request cpu=1.6 were to come in, and kubelet can accommodate, then apiserver won’t apply the Deferred status at T=6 because Deferred patch from kubelet is qualified with the stale spec.containers[0].resources.requests[cpu] = 2. The above sequence would then proceed as below:
```
...
...
T=5: User update #3 requesting cpu = 1.6 via /resize subresource, apiserver patches pod
    spec.containers[0].resources.requests[cpu] = 1.6
    status.Resize[cpu] = 'Proposed'.   <-- apiserver sets this because user changed spec resources (cpu 2 -> 1.6) and spec.resources != status.resources
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=6: Kubelet sees update #2 (request.cpu = 2), possible but not right now, sends patch {spec.containers[0].resources.requests[cpu] = 2, status.Resize = Deferred}
    spec.containers[0].resources.requests[cpu] = 1.6
    status.Resize[cpu] = ‘Proposed’ <-- apiserver drops transition from Proposed to Deferred because Deferred applies to cpu = 2, not the latest proposed cpu = 1.6
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=7: Kubelet sees update #3, accepts it, updates internal checkpoint cpu = 1.6, sends patch {spec.containers[0].resources.requests[cpu] = 1.6, status.Resize = InProgress}, and then invokes CRI.
    spec.containers[0].resources.requests[cpu] = 1.6
    status.Resize[cpu] = 'InProgress'   <-- apiserver sets ‘InProgress’ as {for-atleast-one-ctr spec.resources[cpu] != status.resources[cpu]} and spec in patch matches pod spec.
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=8: Container runtime applied cpu=1.6, CRI returns success, kubelet patches status.resources[cpu] = 1.6
    spec.containers[0].resources.requests[cpu] = 1.6
    status.Resize[cpu] = ''   <-- apiserver clears this because foreach c spec.resources[cpu] == status.resources[cpu]
    status.containerStatuses[0].resources.requests[cpu] = 1.6
```
- The above scheme should also work for multiple container resource updates, as the entire change is either accepted or rejected as an atomic unit, so we don't need per-container Resize.
- Scheduler, resource-quota evaluator would need to use max(podSpec.C[x].Resources, podStatus.CS[x].Resources) to compute node-available for scheduling decisions, RQ evaluation respectively.

**Question: Should we use subresource or mutating admission-controller instead of such specialized interpretation of patch from kubelet at T=4, T=6 above? (Only node can set Resize = InProgress|Deferred|Infeasible)**

API convention considerations - For the above, unless I missed something, I believe the following points still hold true: 
* We are not using status.Resize to store state at any given point or risking it becoming a state-machine.
* If status were lost:
    * apiserver would set Resize = ‘Proposed’ because spec.Resources != status.Resources
    * kubelet would ‘discover’ status.requests & status.limits from checkpoint + CRI, and if spec.Resources != status.Resources, redo canAdmitPod to recompute Resize = InProgress/Deferred/Infeasible.
* If apiserver restarts, nothing should be lost as it would read etcd and get the latest pod state.
* If kubelet restarts at any point, the last checkpoint value + query to CRI will help discover the status.requests & limits values.

Can you please double-check me on this?

Reply:
I commented on the other comment, I will have to think this through to see if we arrived at the same result.  AFK for a bit.

Reply:
> What we lose is the ability to know what the node has actually reserved for the pod when Resize == 'Deferred' or 'Infeasible', and the impact of un-converged max(spec.resources, status.resources), particularly on scheduling. This worries me somewhat.. but the good thing is if we run into real-world problems, we can always just add it later and change scheduler / RQ to always use resourcesAllocated instead of max as we do in the current implementation.

I think I see what you mean.  Specifically, consider a 4 core machine, with one
pod cpu=1.  A resize was requested to cpu=2.

```
t=0: resize in-progress from cpu=1 to cpu=2
  - spec.containers[0].resources.requests[cpu] = 2
  - status.containerStatuses[0].resourcesAllocated[cpu] = 2
  - status.Resize[cpu] = 'InProgress'
  - status.containerStatuses[0].resources.requests[cpu] = 1

t=1: user requests cpu=100
  - apiserver auto-changes status.Resize[cpu] to "Proposed"
  - spec.containers[0].resources.requests[cpu] = 100
  - status.containerStatuses[0].resourcesAllocated[cpu] = 2
  - status.Resize[cpu] = 'Proposed'
  - status.containerStatuses[0].resources.requests[cpu] = 1

t=2: kubelet know it can not satisfy cpu=100
  - spec.containers[0].resources.requests[cpu] = 100
  - status.containerStatuses[0].resourcesAllocated[cpu] = 2
  - status.Resize[cpu] = 'Infeasible'
  - status.containerStatuses[0].resources.requests[cpu] = 1

t=3: resize to cpu=2 completes
  - spec.containers[0].resources.requests[cpu] = 100
  - status.containerStatuses[0].resourcesAllocated[cpu] = 2
  - status.Resize[cpu] = 'Infeasible'
  - status.containerStatuses[0].resources.requests[cpu] = 2
```

Between T=2 and T=3, an observer (scheduler) might be confused.  Without
`resourcesAllocated`, it would assume that 100 is never
going to happen and current = 1.  It could then schedule a cpu=3 pod, which we
KNOW can not proceed.  So it needs `resourcesAllocated` to know the current
goal.  My assertion that it was always the same is wrong.

If that's right, then we need to keep it.  If so, I am forced to ask what value
`Resize` provides.  It clarifies "feasible but not now" from "infleasible", and
it clarifies whether kubelet has actually responded at all.  Still useful, I
think?

> Question: Should we use subresource or mutating admission-controller instead of such specialized interpretation of patch

I think we should do neither.  I don't see that we need to be different or
special here.  I have to drop out for a bit again, but I will come back and
evaluate your status assertions and respond more.

Reply:
Yeah, unless you convince me that you were wrong about resourcesAllocated, I think we can keep it.

Reply:
> Yeah, unless you convince me that you were wrong about resourcesAllocated, I think we can keep it.

Without resourcesAllocated, the probability that scheduler will send a pod destined to be rejected increases based on how big (T=3) - (T=2) gets. But it is not the end of the world. Controller will replace the failed pod with a new one, that will hopefully find a home on try # 2. However, it is conceivable that in a stressed system with lots of pod churn, this will exacerbate to the situation .. hard to say without real-world data.

> If that's right, then we need to keep it. If so, I am forced to ask what value
> Resize provides. It clarifies "feasible but not now" from "infleasible", and
> it clarifies whether kubelet has actually responded at all. Still useful, I
> think?

Yes!! Signal vs. no signal in the Deferred/Infeasible cases. We need to resolve whether to summarize or not - I will jot down both cases in more detail so we can visualize how it plays out.

Reply:
Let's say we went with Resize at container level and look at the following scenario. (No competing/conflicting updates for simplicity)

Say we have a 4 CPU node, and we have a 2-container pod P1 with each container using 1 cpu, and another pod P2 that is using 1 CPU. 

```
T=0: P1 running with total 2 cpus, P2 with 1 CPU.
    P1.spec.containers[0].resources.requests[cpu] = 1
    P1.spec.containers[1].resources.requests[cpu] = 1
    P1.status.containers[0].Resize[cpu] = ''  (no value)
    P1.status.containers[0].Resize[cpu] = ''  (no value)
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
T=1: P1 is updated asking for 3 CPUs more as follows.
    P1.spec.containers[0].resources.requests[cpu] = 3
    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[0].Resize[cpu] = 'Proposed'
    P1.status.containers[0].Resize[cpu] = 'Proposed'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
T=2: Kubelet sees P1 update #1.
{ Computation: P1-new{5}, 5 > NodeCapacity{4} => 'Infeasible' }
    P1.spec.containers[0].resources.requests[cpu] = 3
    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[0].Resize[cpu] = 'Infeasible'  
    P1.status.containers[0].Resize[cpu] = 'Infeasible'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
T=3: P1 is updated asking for 2 CPUs more instead of 3.
    P1.spec.containers[0].resources.requests[cpu] = 3
    P1.spec.containers[1].resources.requests[cpu] = 1
    P1.status.containers[0].Resize[cpu] = 'Proposed'
    P1.status.containers[0].Resize[cpu] = 'Proposed'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
T=2: Kubelet sees P1 update #2.
{ Computation: Sum(P2{1} + P1-new{4}) = 5 > NodeCapacity{4} => 'Deferred' }
    P1.spec.containers[0].resources.requests[cpu] = 3
    P1.spec.containers[1].resources.requests[cpu] = 1
    P1.status.containers[0].Resize[cpu] = 'Deferred'
    P1.status.containers[0].Resize[cpu] = 'Deferred'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
```
Note: At T=2 and T=4, node could accommodate one of the containers but then it how would it decide which one to favor? And whatever it decides, that might still not work for the application as a whole and VPA might need to boot the pod anyways.

This above is because canAdmitPod sums up each pod's total requirements and tests it against what is available in the node - pod level atomicity for evaluating resource requirements. Hence, above we see containers show 'Infeasible' or 'Deferred' in lockstep. So if we summarized it to the pod-level, we would not lose anything.

In either case, we have to nail down a mechanism to resolve competing updates such as kubelet setting Deferred/Infeasible after evaluating an earlier revision after new /resize sets it to Proposed before the kubelet update hits.

Reply:
Regarding competing simultaneous updates, I don't see why we wouldn't proceed with the ones that are feasible and block ones that are not, and if that comes to a "pick one" scenario kubelet can just pick which one it accepts by its own logic?

Consider the case of 2 updates to 2 containers, but NOT initiated at the same time.  Again, copying your timeline:

This is the more obvious case, where we see one good change then one bad
change:

```
T=0: P1 running with total 2 cpus, P2 with 1 CPU.
    P1.spec.containers[0].resources.requests[cpu] = 1
    P1.status.containers[0].Resize[cpu] = ''  (no value)
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

    P1.spec.containers[1].resources.requests[cpu] = 1
    P1.status.containers[1].Resize[cpu] = ''  (no value)
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resourcesAllocated[cpu] = 1

T=1: P1 c[1] is updated asking for 1 CPU more
    P1.spec.containers[0].resources.requests[cpu] = 1
    P1.status.containers[0].Resize[cpu] = ''
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[1].Resize[cpu] = 'Proposed'
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

T=2: Kubelet sees update #1.
{ Computation: P1-new{3}, 3 <= NodeCapacity{4} => 'Feasible' }
{ Computation: P1-new{3} + P2{1} = 4; 4 <= NodeCapacity{4} => 'Feasible' }

    P1.spec.containers[0].resources.requests[cpu] = 1
    P1.status.containers[0].Resize[cpu] = ''
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[1].Resize[cpu] = 'InProgress'
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resourcesAllocated[cpu] = 2

T=3: P1 c[0] is updated asking for 1.5 CPUs more.
    P1.spec.containers[0].resources.requests[cpu] = 2.5
    P1.status.containers[0].Resize[cpu] = 'Proposed'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[1].Resize[cpu] = 'InProgress'
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resourcesAllocated[cpu] = 2

T=4: Kubelet sees update #2.
{ Computation: P1-new{4.5}, 4.5 > NodeCapacity{4} => 'Infeasible' }

    P1.spec.containers[0].resources.requests[cpu] = 2.5
    P1.status.containers[0].Resize[cpu] = 'Infeasible'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[1].Resize[cpu] = 'InProgress'
    P1.status.containerStatuses[1].resources.requests[cpu] = 1
    P1.status.containerStatuses[1].resourcesAllocated[cpu] = 2

T=5: Kubelet finishes update #1.
    P1.spec.containers[0].resources.requests[cpu] = 2.5
    P1.status.containers[0].Resize[cpu] = 'Infeasible'
    P1.status.containerStatuses[0].resources.requests[cpu] = 1
    P1.status.containerStatuses[0].resourcesAllocated[cpu] = 1

    P1.spec.containers[1].resources.requests[cpu] = 2
    P1.status.containers[1].Resize[cpu] = ''
    P1.status.containerStatuses[1].resources.requests[cpu] = 2
    P1.status.containerStatuses[1].resourcesAllocated[cpu] = 2
```

2 interesting points/questions.

1) Is "infeasible" correct for P1 c[0] at T=4?  Strictly speaking it is
feasible, it's just not feasible in conjunction with P1 c[1] at this time.
Would it be more correct to say "Deferred", as if P1 c[1] were external?  My
feel is no - the distinction being "in the same pod".  That said, if c[1] were
to be resized down to 1, c[0] would change from "Infeasible" to "Deferred", and
if c[1] was resized to 0.5, it would become "InProgress".  This sort of breaks
the idea that "Infeasible" is a terminal condition, whereas "Deferred" will be
re-evaluated.

2) Is per-container `resize[]` useful, neutral, or actively harmful?  In this
example it feels like it might be the latter.  I can *mostly* comprehend what
is happening with a single `resize` and per-container `resourcesAllocated`
(which I think is what you were saying).  The single resize REMOVES the
ambiguity above and 'Infeasible' remains a terminal condition (until a
subsequent update happens).  The only thing I lose with single `resize` is the
ability to know that some containers are still in progress while the overall
pod is infeasible.  I'm not sure that matters much

If we flipped the order of updates (so we saw the infeasible change first)
would we even try to accommodate the 2nd change (which is, itself, feasible) as
long as the overall pod was still infeasible?

As an unrelated:

Why these names:
  - containerStatuses[c].resources.requests[r]
  - containerStatuses[c].resourcesAllocated[r]

vs:
  - containerStatuses[c].resources.requests[r]
  - containerStatuses[c].resources.allocated[r]

Reply:
> 
> 2 interesting points/questions.
> 
> 1. Is "infeasible" correct for P1 c[0] at T=4?  Strictly speaking it is
>    feasible, it's just not feasible in conjunction with P1 c[1] at this time.
>    Would it be more correct to say "Deferred", as if P1 c[1] were external?  My
>    feel is no - the distinction being "in the same pod".  That said, if c[1] were
>    to be resized down to 1, c[0] would change from "Infeasible" to "Deferred", and
>    if c[1] was resized to 0.5, it would become "InProgress".  This sort of breaks
>    the idea that "Infeasible" is a terminal condition, whereas "Deferred" will be
>    re-evaluated.
> 2. Is per-container `resize[]` useful, neutral, or actively harmful?  In this
>    example it feels like it might be the latter.  I can _mostly_ comprehend what
>    is happening with a single `resize` and per-container `resourcesAllocated`
>    (which I think is what you were saying).  The single resize REMOVES the
>    ambiguity above and 'Infeasible' remains a terminal condition (until a
>    subsequent update happens).  The only thing I lose with single `resize` is the
>    ability to know that some containers are still in progress while the overall
>    pod is infeasible.  I'm not sure that matters much
> 
> If we flipped the order of updates (so we saw the infeasible change first)
> would we even try to accommodate the 2nd change (which is, itself, feasible) as
> long as the overall pod was still infeasible?
> 

This ambiguity is partly the reason I'm leaning towards having this at PodStatus level instead of per-container level. If we were looking at c[0] alone, it should be 'Deferred', but taking the resource requested by the entire pod it is 'Infeasible'. And if both containers were updated at the same time from 1 -> 2.5 , 1 -> 2, the node can grant one of them but it would have to favor one over the other perhaps by flipping a coin, which is not what the user might want. They may need to end up doing /evict anyways.

Since we are keeping resourcesAllocated, the user has a way to know what was accepted, and is 'InProgress' by the test: spec..resources.requests == status..resourcesAllocated != status.resources.requests.

I don't think it is harmful to have this at per-container level but I'm ignoring a slight gut feeling when I say that.

> As an unrelated:
> 
> Why these names:
> 
> * containerStatuses[c].resources.requests[r]
> * containerStatuses[c].resourcesAllocated[r]
> 
> vs:
> 
> * containerStatuses[c].resources.requests[r]
> * containerStatuses[c].resources.allocated[r]

There are 3 items here:
- containerStatuses[c].resourcesAllocated[r]: The allocation / reservation that node guarantees for the container.
- containerStatuses[c].resources.requests[r]: The reservation enacted by the runtime (applicable to CPU - kernel guarantee that you will get this minimum, we don't have memory reservation in the kernel today, but might have it in the future)
- containerStatuses[c].resources.limits[r]: The limits enforced by kernel. (CPU cap, exceed this memory at risk of OOM kill)

We can define a new type to bring resourcesAllocated under status.resources. Either is fine. Slight lean to how it is currently mainly due to familiarity with it.

Reply:
I buy your argument about pod-scoped `resize` and will push a new commit.

Let's defer the question of `resourcesAllocated` vs `resources.allocated` to the PR.  I have a slight lean toward the nested form, but only slight.